Expand DMS format and fix sec rounding logic

laigyu opened the pull request:
re:#5144

try to let  45°,90°0'0.5" display as 45°,90°0'1"

precision error explain: 
if the sec input is 0.5 ,after calculating it will not equal 0.5 but 0.499999999829015,
my soloution is  round it  by 10e-8 to fix it to 0.5
and do the origin display round 0.5=>1


new format
|35  11  10.1 , 136  49  53.8|| D M SS|
|35  11.168 , 136  49.896||D MM|

zoom center loc always use the user input ,
dont matter how display rounding and show,
always center correct.

laigyu committed:
Expand DMS format and adjust sec to round to .1

re:#5144

Sec 15.56 -> 15.6 , 15.55-> 15.5

Most DMS format parsed by @mapbox/sexagesimal,
add a expandable matcher and two new format to cover some

1ec5 commented on modules/util/units.js:
```diff
@@ -168,3 +171,44 @@ export function decimalCoordinatePair(coord) {
+// return value format : [D, D]  ex:[ 35.1861, 136.83161 ]
+export function dmsMatcher(q) {
+    const matchers = [
+        // D M SS , D M SS  ex: 35 11 10.1 , 136 49 53.8
```
These examples would make for great test cases. Would you mind creating a test/spec/util/units.js that tests this and other edge cases? Thanks!

1ec5 commented on modules/util/units.js:
```diff
@@ -108,31 +108,34 @@ function clamp(x, min, max) {
-        displayCoordinate = displayDegrees +
+    var displayCoordinate;
+    var secToFixedDisplayPrecision = 1;
+    var secFixed = Number(sec.toFixed(secToFixedDisplayPrecision));
```
The previous code rounded the seconds, so this should similarly round the seconds to the nearest tenth instead of flooring it to the lower tenth.

laigyu replied:
It seems tofixed is not a proper way , so I change it to Math.round.
sec is always not precise after calculate from deg, ex: 0.05 after calculate became 0.04999999 , 
so I just round it two time ,try to fix it at the first round,

I will try to write test to test this and dmsMatcher case

1ec5 reviewed (commented):
Thanks! This is looking pretty good at a glance. Perhaps some tests would be a good idea, since there are several edge cases when we do the rounding manually.

laigyu committed:
Fix sec rounding

After calculating ,sec will has precision error,
ex:0.05 after cal -> 0.049999999,
I just use Math.round two time,
plz use 0.04999999 at sec and add 9 to see what happen.

tyrasd commented on modules/util/units.js:
```diff
@@ -106,33 +106,51 @@ function clamp(x, min, max) {
+    var minFloor = Math.floor(min);
+    var sec = (min - minFloor) * 60;
+
+    // console.log(sec);
```
please drop commented out debug code :nerd_face:

laigyu replied:
thanks the review.I didn't consider that. I agree with you , I will change it to no decimal in seconds.

I add D M S because the orgin post want it. and DMM I found it at [google](https://support.google.com/maps/answer/18539?hl=en&co=GENIE.Platform%3DDesktop). I really dont know is common in practice or not.

tyrasd commented on modules/ui/feature_list.js:
```diff
@@ -5,7 +5,7 @@ import * as sexagesimal from '@mapbox/sexagesimal';
 import { presetManager } from '../presets';
 import { t } from '../core/localizer';
-import { dmsCoordinatePair } from '../util/units';
+import { dmsCoordinatePair ,dmsMatcher } from '../util/units';
```
```suggestion
import { dmsCoordinatePair, dmsMatcher } from '../util/units';
```

tyrasd reviewed (commented):
Could you please explain the rationale behind adding the additional decimal precision? IMHO, the resolution of the single minutes is already far sufficient for the intended use case (i.e. navigating to a specified location). 0.1 seconds would only translate to a few additional pixels on the screen… A small negative side effect of showing the decimal is that in some locales (e.g. German) this would introduce the output of a decimal comma in the displayed coordinate which would not be supported as a valid input in the search box (e.g. `12°N, 12°3'4,5"W`).

Are the `D M SS` / `D MM` formats really found in practice? I guess it would not hurt to support it additionally for convenience (e.g. when copy-pasting from spreadsheets/csv files with dedicated columns for each coordinate part). I'd agree that it would be required to add some unit tests for these routines. And maybe also add support for it with N/S and E/W suffixes in that format (e.g. `45 0 18 N, 90 19 0 W`)?

laigyu committed:
Round sec to the nearest integer

laigyu committed:
adjust space

laigyu committed:
fix typo

tyrasd committed:
simplify inline comment

tyrasd committed:
spell checker adjustments

tyrasd committed:
Merge branch 'develop' into Adjust-DMS-display-and-add-new-format

tyrasd committed:
add basic tests

tyrasd committed:
changelog: support for coords in DMS format in search bar

example: `35 11 10.1 , 136 49 53.8` (D M SS format), or `35 11.168 , 136 49.896` (D MM format)

laigyu committed:
add more tests

tyrasd commented:
Thanks for the additional test cases!
